libcudacxx/docs/extended_api/memory_resource.md

```diff
@@ -0,0 +1,220 @@
+## memory_resource
+
+The `cuda::memory_resource` header provides two main features:
+* The `cuda::resource` and `cuda::resource_with` concepts that provide proper constraints for arbitrary memory resoures
+* The `cuda::resource_ref` resource wrapper as a type erased memory resource, that allows for simple efficient interfaces
+
+These features are an evolution of `std::pmr::memory_resource` that was introduced in C++17. While `std::pmr::memory_resource` provides a polymorphic memory resource that can be adopted through inheritance, it is not properly suited for heterogeneous systems. With the current design it ranges from cumbersome to impossible to verify whether a memory resource provides allocations that are e.g. accessible on device, or whether it can utilize other allocation mechanisms.
+
+### Properties
+
+With C++ we want the type system to verify statically known properties of our code. If a memory resource provides allocations that are device accessible, we do not want to rely on runtime asserts to check whether we can indeed call a device function with that memory. Therefore, we are introducing a property system that is similar to rust traits.
+
+Let's assume we want to tell the compiler that memory provided by `my_memory_resource` is indeed device accessible. For that we introduce a tag type:
+```c++
+namespace cuda::mr {
+    struct device_accessible{};
+}
+```
+
+We can utilize a free function `get_property` to mark `my_memory_resource` as providing device accessible memory
+```c++
+struct my_memory_resource {
+    friend constexpr void get_property(const my_memory_resource&, cuda::mr::device_accessible) noexcept {}
+};
+```
+
+A library can now constrain their interface with `cuda::mr::has_property` to verify that the passed memory resource is providing the right kind of memory (See [here](https://godbolt.org/z/MTTd6dqe1) for a minimal godbolt example)
+```c++
+template<class MemoryResource>
+    requires cuda::mr::has_property<MemoryResource, cuda::mr::device_accessible>
+void function_that_dispatches_to_device(MemoryResource& resource);
+```
+
+For now libcu++ provides two standard properties `cuda::mr::device_accessible` and `cuda::mr::host_accessible`. We envision to add a few more as the library and the hardware capabilities evolve. However, a user library is free to define as many properties as they want to fully cover their respective API surface.
```

Comment:
Not sure if we want to call these "standard" properties; we want to convey that they are meant to be commonly used, but since the word "standard" has quite the connotations, we probably should find another word for this.